src/models/base_model.py

```diff
@@ -0,0 +1,218 @@
+    internal nodes in a model by node name. This works quite well in eager Python but needs
+    redesign for torcscript.
+    """
+
+    def __init__(self, hooks: List[Hook], named_modules: Generator[Tuple[str, nn.Module]]):
+        """Init FeatureHooks class.
+        
+        Args:
+            hooks: Hooks to be registered.
+            named_modules: Result of self.named_modules() function call.
+        """
+        # setup feature hooks
+        modules = {k: v for k, v in named_modules}
+        for i, hook in enumerate(hooks):
+            hook_name = hook.module_name
+            module = modules[hook_name]
+            hook_fn = partial(self._collect_hook_features, hook_name)
+            # register hooks
+            if hook.hook_type == HookType.FORWARD_PRE:
+                module.register_forward_pre_hook(hook_fn)
+            else:
+                module.register_forward_hook(hook_fn)
+           
+        self._feature_outputs = defaultdict(OrderedDict)
+
+    def _collect_hook_features(self, hook_name: str, *args):
+        """Collect hooks features in self._hook_features.
+        
+        Args:
+            hook_name: Name of saved hooked module.
+        """
+        # tensor we want is last argument, output for fwd, input for fwd_pre
+        x = args[-1]  
+        if isinstance(x, (tuple, list)) and len(x) == 1:
+            # unwrap input tuple
+            x = x[0]  
+            device = x.device
+        else:
+            device = x[0].device
+        self._hook_features[device][hook_name] = x
+
+    def get_features(self, device: torch.device) -> Dict[str, torch.tensor]:
+        """Return hooks features.
+
+        Args:
+            device: Saved hooks device.
+
+        Returns:
+            output: Dict of module hook name two its output tensor.
+        """
+        output = self._hook_features[device]
+        # clear after reading
+        self._hook_features[device] = OrderedDict()  
+        return output
+
+
+class BaseModel(nn.Module, ABC):
+    """Base model for all TorchOk Models - Backbone, Neck, Pooling and Head.
+
+    This class supports adding feature hooks to some model layers.
+    Class has feature info list for every hook.
+    To create hooks, method self.get_features_info() must be rewrited!
+    Also class contains method get_output_channels wich returns forward pass and hooks output channels. 
+    """
+    def __init__(self):
+        """Inits BaseModel class and it's hooks.
+
+        Raises:
+            ValueError: If rewrited self.get_features_info() method return not FeatureInfo list.
+        """
+        super().__init__()
+
+        # Create feature_info list with hooks
+        self._features_info, self._feature_hooks = self._create_hooks()
+
+
+    def get_features_info(self) -> List[FeatureInfo]:
+        """Method for initialize Hooks.
+        
+        It must be rewrite in inherited classes, if Module would need hooks.
```

Comment:
It should be overriden in an inherited class if a user is expected to have access to feature hooks of the implemented Module